compiler/codegen/src/compile.rs

```diff
@@ -2877,19 +2905,29 @@ impl Compiler {
         self.compile_expression(&generators[0].iter)?;
 
         // Get iterator / turn item into an iterator
-        if is_async {
+        if has_async_gen {
             emit!(self, Instruction::GetAIter);
         } else {
             emit!(self, Instruction::GetIter);
         };
 
         // Call just created <listcomp> function:
         emit!(self, Instruction::CallFunctionPositional { nargs: 1 });
-        if is_async {
+        if has_async_gen {
+            emit!(self, Instruction::GetAwaitable);
+            self.emit_load_const(ConstantData::None);
+            emit!(self, Instruction::YieldFrom);
+        }
+
+        if is_async && comprehension_type != ComprehensionType::Generator && !has_async_gen {
+            // async, but not a generator and not an async for
+            // in this case, we end up with an awaitable
+            // that evaluates to the list/set/dict, so here we add an await
             emit!(self, Instruction::GetAwaitable);
             self.emit_load_const(ConstantData::None);
             emit!(self, Instruction::YieldFrom);
```

Comment:
Can the both block be true to insert the same instructions twice? And is it correct?

Reply:
I don't think so - one requires `has_async_gen` to be true, but if `has_async_gen` is true, then we have `... && !has_async_gen` and anything && false is false. So they are exclusive. I changed the logic now though to have to booleans set at the beginning for the various cases instead of having logic spread throughout the function so now there is only one block here.